File: app/src/main/java/me/srikavin/quiz/repository/Repository.java
package me.srikavin.quiz.repository;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

public abstract class Repository {
    protected abstract static class ResponseHandler<E, T> {
        @WorkerThread
        public abstract void handle(@Nullable T t);

        @WorkerThread
        public abstract void handleMultiple(@Nullable List<T> t);

        @WorkerThread
        public abstract void handleErrors(@NonNull E[] errors);
    }
}

File: app/src/main/java/me/srikavin/quiz/repository/Repository.kt
package me.srikavin.quiz.repository

import androidx.annotation.WorkerThread

/**
 * Used to fetch data from local or online sources
 */
abstract class Repository {
    /**
     * Used to handle callbacks for repository events
     */
    abstract class ResponseHandler<E, T> {
        /**
         * Handles a result providing a single item.
         *
         * @param t May be null. Contains the specified object
         */
        @WorkerThread
        abstract fun handle(t: T?)

        /**
         * Handles a result providing multiple items
         */
        @WorkerThread
        abstract fun handleMultiple(t: List<T>)

        /**
         * Handles an error that occurred as a result of calling a repository function
         */
        @WorkerThread
        abstract fun handleErrors(vararg errors: E)
    }
}
